config/envoyconfig/listeners.go

```diff
@@ -668,42 +668,47 @@ func (b *Builder) buildDownstreamTLSContext(ctx context.Context,
 	}
 
 	envoyCert := b.envoyTLSCertificateFromGoTLSCertificate(ctx, cert)
-	return &envoy_extensions_transport_sockets_tls_v3.DownstreamTlsContext{
+	dtc := &envoy_extensions_transport_sockets_tls_v3.DownstreamTlsContext{
 		CommonTlsContext: &envoy_extensions_transport_sockets_tls_v3.CommonTlsContext{
-			TlsParams:             tlsParams,
-			TlsCertificates:       []*envoy_extensions_transport_sockets_tls_v3.TlsCertificate{envoyCert},
-			AlpnProtocols:         alpnProtocols,
-			ValidationContextType: b.buildDownstreamValidationContext(ctx, cfg, domain),
+			TlsParams:       tlsParams,
+			TlsCertificates: []*envoy_extensions_transport_sockets_tls_v3.TlsCertificate{envoyCert},
+			AlpnProtocols:   alpnProtocols,
 		},
 	}
+	if clientCA := clientCAForDomain(cfg, domain); len(clientCA) > 0 {
+		dtc.CommonTlsContext.ValidationContextType = b.buildDownstreamValidationContext(ctx, cfg, clientCA)
+		dtc.RequireClientCertificate = wrapperspb.Bool(true)
+	}
+	return dtc
 }
 
-func (b *Builder) buildDownstreamValidationContext(ctx context.Context,
-	cfg *config.Config,
-	domain string,
-) *envoy_extensions_transport_sockets_tls_v3.CommonTlsContext_ValidationContext {
-	needsClientCert := false
-
-	if ca, _ := cfg.Options.GetClientCA(); len(ca) > 0 {
-		needsClientCert = true
-	}
-	if !needsClientCert {
-		for _, p := range getPoliciesForDomain(cfg.Options, domain) {
-			if p.TLSDownstreamClientCA != "" {
-				needsClientCert = true
-				break
-			}
+// clientCAForDomain returns a bundle of all per-route client CAs configured
+// for the given domain, or else the globally configured client CA.
+func clientCAForDomain(cfg *config.Config, domain string) []byte {
+	var bundle []byte
+	for _, p := range getPoliciesForDomain(cfg.Options, domain) {
+		if p.TLSDownstreamClientCA == "" {
```

Comment:
there is also `TLSDownstreamClientCAFile` that need be accounted for.

Reply:
At first I thought so as well, but then I noticed that `Policy.Validate()` will populate `TLSDownstreamClientCA` from `TLSDownstreamClientCAFile`: https://github.com/pomerium/pomerium/blob/0d34e3d00c16d0e951e6e18f1992531cc10c96c6/config/policy.go#L501

Is it safe to assume that `Validate()` will have been called on each `Policy` before we get to the main `BuildListeners()` method?

Reply:
indeed. in general, Validate() function shouldn't really have side effects, so I try to never rely on such magic behaviour. But we may do so here.